cpp/unreal_plugins/SpEngine/Source/SpEngine/WorkQueue.h

```diff
@@ -0,0 +1,133 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#pragma once
+
+#include <concepts>    // std::same_as
+#include <cmath>       // std::nan
+#include <future>
+#include <map>
+#include <mutex>
+#include <string>
+#include <utility>     // std::forward, std::move
+#include <vector>
+
+#include <SpCore/Assert.h>
+#include <SpCore/Boost.h>
+
+template <typename TFunc, typename TReturn, typename... TArgs>
+concept CCallable = requires(TFunc func, TArgs... args) {
+        { func(args...) } -> std::same_as<TReturn>;
+};
+
+
+template <typename TClass>
+struct FuncInfo : public FuncInfo<decltype(&TClass::operator())> {};
+
+template <typename TClass, typename TReturn, typename... TArgs>
+struct FuncInfo<TReturn(TClass::*)(TArgs...)> : public FuncInfo<TReturn(*)(TArgs...)> {};
+
+template <typename TClass, typename TReturn, typename... TArgs>
+struct FuncInfo<TReturn(TClass::*)(TArgs...) const> : public FuncInfo<TReturn(*)(TArgs...)> {};
+
+template <typename TReturn, typename... TArgs>
+struct FuncInfo<TReturn(*)(TArgs...)> {};
+
+
+// The purpose of this class is to provide a copy-constructible type that derives from std::packaged_task. This is necessary 
+// because boost::asio requires that task objects are copy-constructible, but std::packaged_task is not copy-constructible.
+template <typename TReturn>
```

Comment:
I was able to remove `FuncInfo` from `scheduleAndExecuteTaskBlocking` but not from `wrapFuncToExecuteInWorkQueueBlocking`. The difference is that `scheduleAndExecuteTaskBlocking` only needs to refer to `TFunc` and `TReturn`, but `wrapFuncToExecuteInWorkQueueBlocking` needs `TFunc` and `TReturn` and `TArgs`.

Reply:
Also I don't think the `createCopyConstructiblePackagedTask()` function is necessary any more.

Reply:
I agree, looks like we will need `FuncInfo` after all. I simplified this code.

Reply:
I like your approach better. Indeed, that's how I originally specified it in the design doc. I just didn't want to deal with the `WorkQueue` class when I was experimenting with the snippet I posted above.